### PR TITLE
Expand NVTX usage

### PR DESCRIPTION
**Pull request by ptrendx**
## Description ##
This PR improves the experience when profiling MXNet under nvprof or Nsight Systems, by expanding NVTX functionality.
It was used successfully for some time in the NVIDIA NGC container.

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- [x] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] Added C API for NVTX ranges push/pop
- [x] Added C API to start/stop profiling via nvprof/NSight Systems
- [x] Added Python API corresponding to the C API
- [x] Added automatic ranges to MXNet operators

### CODE REVIEW

**Comment by szha** [comment]
@mxnet-bot run ci [centos-cpu, miscellaneous]

**Comment by szha** [comment]
@ptrendx still WIP?

**Comment by ptrendx** [comment]
Yes, I will return back to it sometime next week probably (@DickJC123 is currently merging 1.7 into our codebase and cherry picking stuff after that merge will be easier).

**Comment by szha** [comment]
What's the status for this change?

**Comment by ptrendx** [comment]
It is currently on hold - I got sidetracked to other things and am not sure exactly when I will have time to finish it - I still intend to do that, though.

**Comment by ptrendx** [comment]
@mxnet-bot run ci [unix-cpu]

File: python/mxnet/cuda_utils.py
Comment (DickJC123) [inline python/mxnet/cuda_utils.py:21]: I don't enjoy suggesting a code repetition, but isn't the following an improvement over letting python users have to make their own colors:
```
class Nvtx:
    # Palette of colors
    RED = 0xFF0000
    GREEN = 0x00FF00
    BLUE = 0x0000FF
    YELLOW = 0xB58900
    ORANGE  = 0xCB4B16
    RED = 0xDC322F
    MAGENTA = 0xD33682
    VIOLET = 0x6C71C4
    BLUE1   = 0x268BD2
    CYAN    = 0x2AA198
    GREEN  = 0x859900

def nvtx_range_push(name, color=Nvtx.ORANGE):
```

File: src/common/nvtx.h
Comment (DickJC123) [inline src/common/nvtx.h:69]: Do we know if the nvtxRangePushEx(&att) makes a copy of att.message.ascii?  I suspect not, in which case I'm worried about the lifetime of range_name.c_str().

gpuRangeStart() is called with a const char *, so won't that require making a short-lived temporary string (with a copy of the char array) in the caller's environment?

File: src/common/nvtx.h
Comment (DickJC123) [inline src/common/nvtx.h:1]: Should this new file be ./src/common/cuda/nvtx.h ?

File: src/common/nvtx.h
Comment (ptrendx) [reply src/common/nvtx.h:69]: Not sure I your concern - nvtxRangePush is not an async call. It just puts the name and a timestamp into the database and that's it (and yes, it copies the string in the process).

File: src/common/nvtx.h
Comment (ptrendx) [reply src/common/nvtx.h:1]: Could be, will change that.

File: src/common/nvtx.h
Comment (DickJC123) [reply src/common/nvtx.h:69]: So if nvtxRangePush copies the att.message.ascii string (as opposed to the pointer), that answers my primary concern here.

I'm left with one nit then, why don't you have `const char * range_name` above, which avoids the needless string copy and keeps the type consistent between the callers (in c_api.cc and threaded_engine_perdevice.cc) and what att.message.ascii wants?

Comment (ptrendx) [reply src/common/nvtx.h:69]: It is a reference to a string, so there is no copying here (`c_str()` does not copy either). I believe in the threaded engine the nvtx_name is a string anyway, so that is also not doing any copies.

File: src/common/cuda/nvtx.h
Comment (DickJC123) [inline src/common/cuda/nvtx.h:90]:
```suggestion
  static uint32_t nameToColor(const std::string& range_name) {
```

File: src/common/nvtx.h
Comment (DickJC123) [reply src/common/nvtx.h]: In threaded engine, the following does a copy that could be avoided if everything was handled as a `const char *`:
```
auto nvtx_name = opr_block->opr->opr_name != "" ? opr_block->opr->opr_name : "Op";
```
so perhaps like:
```
auto nvtx_name = opr_block->opr->opr_name.size() ? opr_block->opr->opr_name.c_str() : "Op";
```
It's a nit as I say, so I'm OK leaving it as: `const std::string& range_name`.  Given that though, the two overloaded nameToColor() signatures should be consistent in their use of std::string (see an additional suggestion).

File: python/mxnet/cuda_utils.py
Comment (DickJC123) [reply python/mxnet/cuda_utils.py:21]: If you're not fond of the code repetition, at least the following smaller edit explains the magic number:
```
def nvtx_range_push(name, color=0xCB4B16):    # default color is orange
```

File: src/common/cuda/nvtx.h
Comment (ptrendx) [reply src/common/cuda/nvtx.h:90]: :+1:

File: python/mxnet/cuda_utils.py
Comment (ptrendx) [reply python/mxnet/cuda_utils.py:21]: I am refactoring this to be nicer and more Pythonic. Will send an updated version tomorrow, once I test it.

Comment (ptrendx) [reply python/mxnet/cuda_utils.py:21]: The problem with this is mainly that the standard hash in Python is not deterministic, which would make colors change on subsequent runs. Using a real hashing function like the SHA from Python's hashlib on the other hand seems to be quite a heavy hammer (and most probably introduce additional overheads, which we especially do not want to do when profiling the CPU part of the training).

Comment (DickJC123) [reply python/mxnet/cuda_utils.py]: What I tried to suggest was that the 'None' color would be conveyed to the backend (via -1 or alternate API with no color arg), and the backend would assign the color with the existing backend nameToColor() routine.

**Comment by ptrendx** [comment]
@mxnet-bot run ci [unix-cpu, website]

**Comment by ptrendx** [comment]
@mxnet-bot run ci [unix-cpu, website]

**Comment by ptrendx** [comment]
@mxnet-bot run ci [unix-cpu]

**Review by DickJC123 — APPROVED** [review]
Thanks for responding to my comments.  LGTM.

**Pull request merged**
